lib/connection_controller.go

```diff
@@ -0,0 +1,330 @@
+	if err != nil {
+		return err
+	}
+	return cc.rnManager.CreateNonValidatorPersistentOutboundConnection(netAddr)
+}
+
+func (cc *ConnectionController) CreateNonValidatorOutboundConnection(ipStr string) (_err error) {
+	netAddr, err := cc.ConvertIPStringToNetAddress(ipStr)
+	if err != nil {
+		return err
+	}
+	return cc.rnManager.CreateNonValidatorOutboundConnection(netAddr)
+}
+
+func (cc *ConnectionController) SetTargetOutboundPeers(numPeers uint32) {
+	cc.targetOutboundPeers = numPeers
+}
+
+func (cc *ConnectionController) enoughNonValidatorInboundConnections() bool {
+	return uint32(cc.rnManager.GetNonValidatorInboundIndex().Count()) >= cc.maxInboundPeers
+}
+
+func (cc *ConnectionController) enoughNonValidatorOutboundConnections() bool {
+	return uint32(cc.rnManager.GetNonValidatorOutboundIndex().Count()) >= cc.targetOutboundPeers
+}
+
+// processInboundConnection is called when a new inbound connection is established. It is responsible for validating
+// the connection and creating a RemoteNode from it.
+func (cc *ConnectionController) processInboundConnection(conn Connection) (*RemoteNode, error) {
+	var ic *inboundConnection
+	var ok bool
+	if ic, ok = conn.(*inboundConnection); !ok {
+		return nil, fmt.Errorf("ConnectionController.handleInboundConnection: Connection is not an inboundConnection")
+	}
+
+	// Reject the peer if we have too many inbound connections already.
+	if cc.enoughNonValidatorInboundConnections() {
+		return nil, fmt.Errorf("ConnectionController.handleInboundConnection: Rejecting INBOUND peer (%s) due to max "+
+			"inbound peers (%d) hit", ic.connection.RemoteAddr().String(), cc.maxInboundPeers)
+	}
+
+	// If we want to limit inbound connections to one per IP address, check to make sure this address isn't already connected.
+	if cc.limitOneInboundConnectionPerIP &&
+		cc.isDuplicateInboundIPAddress(ic.connection.RemoteAddr()) {
+
+		return nil, fmt.Errorf("ConnectionController.handleInboundConnection: Rejecting INBOUND peer (%s) due to "+
+			"already having an inbound connection from the same IP with limit_one_inbound_connection_per_ip set",
+			ic.connection.RemoteAddr().String())
+	}
+
+	na, err := cc.ConvertIPStringToNetAddress(ic.connection.RemoteAddr().String())
+	if err != nil {
+		return nil, errors.Wrapf(err, "ConnectionController.handleInboundConnection: Problem calling "+
+			"ConvertIPStringToNetAddress for addr: (%s)", ic.connection.RemoteAddr().String())
+	}
+
+	remoteNode, err := cc.rnManager.AttachInboundConnection(ic.connection, na)
+	if remoteNode == nil || err != nil {
+		return nil, errors.Wrapf(err, "ConnectionController.handleInboundConnection: Problem calling "+
+			"AttachInboundConnection for addr: (%s)", ic.connection.RemoteAddr().String())
+	}
+
+	return remoteNode, nil
+}
+
+// processOutboundConnection is called when a new outbound connection is established. It is responsible for validating
+// the connection and creating a RemoteNode from it.
```

Comment:
Ditto on it being unlcear what it means that a new outbound connection has been established. Has the connection been dialed and formed? Has the handshake process started? Please clarify at what stage of the connection lifecycle this is called